gremlin-go/driver/graphBinary.go

```diff
@@ -1282,6 +1284,78 @@ func metricsReader(data *[]byte, i *int) (interface{}, error) {
 	return metrics, nil
 }
 
+type JanusgraphRelationIdentifier struct {
+	OutVertexIdLong   int64
+	OutVertexIdString string
+	TypeId            int64
+	RelationId        int64
+	InVertexIdLong    int64
+	InVertexIdString  string
+}
+
+func janusgraphRelationIdentifierReader(data *[]byte, i *int) (interface{}, error) {
+	const (
+		relationIdentifierType uint32 = 0x1001
+		longMarker             byte   = 0
+		stringMarker           byte   = 1
+	)
+
+	r := new(JanusgraphRelationIdentifier)
+
+	// expect type code
+	customDataTyp := readUint32Safe(data, i)
+	if customDataTyp != relationIdentifierType {
+		return nil, newError(err0408GetSerializerToReadUnknownTypeError, customDataTyp)
+	}
+
+	// value flag, expect this to be non-nullable
+	if readByteSafe(data, i) != valueFlagNone {
+		return nil, newError(err0405ReadValueInvalidNullInputError)
+	}
+
+	// outvertexid
+	if readByteSafe(data, i) == longMarker {
+		r.OutVertexIdLong = readLongSafe(data, i)
+	} else {
+		vertexId, err := readString(data, i)
+		if err != nil {
+			return nil, err
+		}
+		r.OutVertexIdString = vertexId.(string)
+	}
+
+	r.TypeId = readLongSafe(data, i)
+	r.RelationId = readLongSafe(data, i)
+
+	// invertexid
+	if readByteSafe(data, i) == longMarker {
+		r.InVertexIdLong = readLongSafe(data, i)
+	} else {
+		vertexId, err := readString(data, i)
+		if err != nil {
+			return nil, err
+		}
+		r.InVertexIdString = vertexId.(string)
+	}
+
+	return r, nil
+}
+
+// {name}{type specific payload}
+func customTypeReader(data *[]byte, i *int) (interface{}, error) {
+	// type name
+	*i = *i - 1
+	customTypeName, err := readString(data, i)
+	if err != nil {
+		return nil, err
+	}
+	deserializer, ok := customDeserializers[customTypeName.(string)]
```

Comment:
For custom types, it looks like we write the name of the type, and then the remaining bytes are specific to the implementation. So it could make sense to key off the name, and let the custom reader handle the rest.

Does this make sense?

Reply:
there is example of implementation of custom serializer for Java, looks like a good starting point
https://github.com/apache/tinkerpop/blob/master/gremlin-util/src/test/java/org/apache/tinkerpop/gremlin/util/ser/binary/types/sample/SamplePersonSerializer.java